### PR TITLE
iOS AutoFill Functionality

### PR DESCRIPTION
**Pull request by se1exin**
This PR introduces native AutoFill functionality for iOS, as discussed in #125 

**Overview**

- Users must manually enable AutoFill and Buttercup in the Passwords section of iOS Settings.
- Buttercup will replicate user credentials into the iOS AutoFill credential store everytime an Archive is opened/updated. iOS will make these available directly in the QuickBar above the keyboard for fastest possible login.
- The user may enter Buttercup via the AutoFill functionality to further search for credentials.
- Buttercup AutoFill will attempt to automatically unlock all touch enabled Archives on load.
- Non-touch enabled Archives can still be accessed via manual master password entry.
- The user can only access the Archive List and Search page. Tapping an Archive goes to the Search page with Search restricted to that Archive.
- Search Results mimic the Buttercup Browser Extension behaviour. With domain related results shown initially and when search has not matching results.
- Tapping a credential will complete the autofill process.
- The user can cancel the process at any time via the 'Cancel' button in header right on both pages.

**Significant changes**

- AsyncStorage replaced with `react-native-secure-storage`. Includes auto-migration for existing users in base `App` component.
- Additional RN Entry Point created - `index.ios.autofill.js`.
- Some top-level containers/components replicated to remove unneeded/incompatible functionality. See `source/autofill/**`
- Additional Redux Store - `autofill`. Used to flag/check if the app is running in the context of the AutoFill extension
- Update to the Xcode Project entitlements. May need some updates in the Apple Dev Account, but Automatic Signing in Xcode should take care of it.

I've tested quite a lot but there may still be bugs, would great to have some other people test it out.

### CODE REVIEW

**Comment by perry-mitchell** [comment]
omfg

**Comment by perry-mitchell** [comment]
@se1exin This is incredible! Thanks so much for pushing this through. Amazing to see a PR for it already.

By the way, do you know if the release process changes much? Do we need to release the extension separately or does it just bundle?

**Comment by sallar** [comment]
@se1exin thanks a lot! extraordinary job :) 
this will take a while to review though haha

File: source/library/buttercup.js [hunk 1]
```diff
@@ -98,7 +99,8 @@ export function getArchiveEncryptedContent(archive, credentials) {
 
 export function getSharedArchiveManager() {
     if (__sharedManager === null) {
-        __sharedManager = new ArchiveManager(new AsyncStorageInterface());
+        // __sharedManager = new ArchiveManager(new AsyncStorageInterface());
```
Comment (perry-mitchell) [inline source/library/buttercup.js]: I guess this can be removed?

File: source/library/buttercup.js [hunk 2]
```diff
@@ -6,7 +6,8 @@ import {
     Credentials,
     entryFacade
 } from "./buttercupCore.js";
-import AsyncStorageInterface from "../compat/AsyncStorageInterface.js";
+// import AsyncStorageInterface from "../compat/AsyncStorageInterface.js";
```
Comment (perry-mitchell) [inline source/library/buttercup.js]: I guess this can also be removed?

**Comment by perry-mitchell** [comment]
We'll probably merge this to a mid-way branch for final testing before going to master. We should also endeavour to strip the Facebook comments form all files (new and old): #134

File: ios/BCAutoFillExtension/Info.plist [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	<key>CFBundleInfoDictionaryVersion</key>
+	<string>6.0</string>
+	<key>CFBundleName</key>
+	<string>$(PRODUCT_NAME)</string>
+	<key>CFBundlePackageType</key>
+	<string>XPC!</string>
+	<key>CFBundleShortVersionString</key>
+	<string>1.0</string>
+	<key>CFBundleVersion</key>
+	<string>1</string>
```
[the remark this answers is not in the document]
Comment (se1exin) [reply ios/BCAutoFillExtension/Info.plist]: Build script added to BCAutoFillExtension target that sync's the Extension Build numbers with the main app's. See commit 908fdcf1e8aa10f5c716fe73d028dbc6ed26d5e9

File: source/library/buttercup.js (hunk 1 above)
Comment (se1exin) [reply source/library/buttercup.js]: Removed in 908fdcf1e8aa10f5c716fe73d028dbc6ed26d5e9

File: source/library/buttercup.js (hunk 2 above)
Comment (se1exin) [reply source/library/buttercup.js]: Removed in 908fdcf1e8aa10f5c716fe73d028dbc6ed26d5e9

**Comment by se1exin** [comment]
Thanks for the positive feedback @perry-mitchell and @sallar!

It turned out to be a bit bigger of a job than I first expected, but much learnt about iOS AuthenticationFramework along the way 😅. Let me know if you need any changes or explanations for anything 👍

> Do we need to release the extension separately or does it just bundle?

It just bundles from the main app's build process, so you shouldn't have to change the release process at all. In 908fdcf1e8aa10f5c716fe73d028dbc6ed26d5e9 I've also added a Build Script in the Extension Build Phases that ensures the Extension version numbers are the same as the main app. It's rumoured on Stack Overflow that Apple actually requires these to be the same for review.

> We should also endeavour to strip the Facebook comments form all files (new and old)

I've attempted to do this for the iOS code files in 908fdcf1e8aa10f5c716fe73d028dbc6ed26d5e9. Wasn't sure what the Copyright should be changed to so just went with `Buttercup` and added reference to GPLv3 where Facebook had previously mentioned MIT. Also updated the organisation in the Xcode project so that future autogenerated copyright comments say Buttercup instead of Facebook 👍.

**Comment by perry-mitchell** [comment]
Thanks @se1exin! I appreciate the trouble you went to to make this easy to deploy. I'll merge now into an intermediary branch for some testing before we go to master.

**Comment by se1exin** [comment]
@perry-mitchell Awesome! Looking forward to having this in the official build so I can start using it on my day-to-day phone!

**Pull request merged**